api/v1alpha1/adminnetworkpolicy_types.go

```diff
@@ -0,0 +1,309 @@
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// All fields in this package are required unless Explicitly marked optional
+// +kubebuilder:validation:Required
+package v1alpha1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// AdminNetworkPolicy is the Schema for the adminnetworkpolicies API
+type AdminNetworkPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired behavior of AdminNetworkPolicy.
+	Spec AdminNetworkPolicySpec `json:"spec,omitempty"`
+
+	// ANPStatus is the status to be reported by the implementation, this is not
+	// standardized in alpha and consumers should report what they see fit in
+	// relation to their AdminNetworkPolicy implementation
+	// +optional
+	Status AdminNetworkPolicyStatus `json:"status,omitempty"`
+}
+
+// AdminNetworkPolicyStatus defines the observed state of AdminNetworkPolicy
+type AdminNetworkPolicyStatus struct {
+	Conditions []metav1.Condition `json:"status"`
+}
+
+// AdminNetworkPolicySubject defines what objects the policy selects.
+// Exactly one of the `NamespaceSelector` or `NamespaceAndPodSelector` pointers
+// should be set.
+type AdminNetworkPolicySubject struct {
+	// +optional
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector"`
+	// +optional
+	NamespaceAndPodSelector *NamespacedPodSubject `json:"namespaceAndPodSelector"`
+}
+
+// NamespacedPodSubject allows the user to select a given set of pod(s) in
+// selected namespace(s)
+type NamespacedPodSubject struct {
+	// This field follows standard label selector semantics; if present but empty,
+	// it selects all Namespaces.
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector"`
+
+	// Used to explicitly select pods within a namespace; if present but empty,
+	// it selects all Pods.
+	PodSelector *metav1.LabelSelector `json:"podSelector"`
+}
+
+// AdminNetworkPolicySpec defines the desired state of AdminNetworkPolicy
+type AdminNetworkPolicySpec struct {
+	// Priority is an int32 value bound to 0 - 1000, the lowest positive priority,
+	// "1" corresponds to the highest importance, while higher priorities have
+	// lower importance. An ANP with a priority of "0" will be evaluated after all
+	// positive priority AdminNetworkPolicies and standard NetworkPolicies.
+	// This field is NOT optional.
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=1000
+	Priority *int32 `json:"priority"`
```

Comment:
This one we decided on the KEP should be a pointer AND required in order to prevent defaulting to 0, meaning "execute after namespaced network policy"

IMO the way we represent ordering with respect to namespaced policies still needs work, though.

Reply:
With the addition of a `baselineAdminNetworkPolicy` object this no longer needs to be a pointer.